src/lib/y2storage/proposal/devices_planner.rb

```diff
@@ -35,215 +31,33 @@ module Proposal
     class DevicesPlanner
       include Yast::Logger
 
-      attr_accessor :settings
+      STRATEGIES = {
+        legacy: DevicesPlannerStrategies::Legacy,
+        ng:     DevicesPlannerStrategies::NG
+      }
 
-      MIN_SWAP_SIZE = DiskSize.MiB(512)
-      MAX_SWAP_SIZE = DiskSize.GiB(2) # This is also DEFAULT_SWAP_SIZE
-      SWAP_WEIGHT = 100 # Importance of swap if low free disk space
+      attr_accessor :settings
 
       def initialize(settings, devicegraph)
         @settings = settings
         @devicegraph = devicegraph
+        strategy = @settings.format
+        @strategy = STRATEGIES[strategy].new(settings, devicegraph)
+      rescue
```

Comment:
Here, an exception could be raised for different reasons (unknown strategy, failure in new, etc). Maybe, we should check specifically the correctness of the strategy:

```
def initialize(settings, devicegraph)
  @settings = settings
  @devicegraph = devicegraph
  @strategy = strategy
end

def strategy
  return @strategy unless @strategy.nil?

  strategy_class = 
    case @settings.format
    when :legacy
      DevicesPlannerStrategies::Legacy
    when :ng
      DevicesPlannerStrategies::Ng
    else
      err_msg = "Unsupported device planner strategy :#{@settings.format}"
      log.error err_msg
      raise DevicesPlannedStrategyError, err_msg
    end

  @strategy = strategy_class.new(@settings, @devicegraph)
end
```

Reply:
In this case, the constant `DevicesPlanner::STRATEGIES` would not be necessary.

Reply:
I'm not convinced. Having different errors for all kind of things provides a useless level of detail. In the end you'll just check for an error and not distinguish several cases while setting a strategy.

Is there anything bad with having a `DevicesPlanner::STRATEGIES` constant?

Reply:
Well, I am not against about using `STRATEGIES` or raise `ArgumentError`, but what I do not see correct is to raise always the same message error independently of the exception. For example, an exception could happen in `@settings.format` or in `STRATEGIES[strategy].new` even when the `STRATEGIES[strategy]` is known.

So, I agree with you propose, but maybe raising the exception only when the strategy is not known:

```
def initialize(settings, devicegraph)
  @settings = settings
  @devicegraph = devicegraph
  strategy = @settings.format
  if STRATEGIES[strategy]
    @strategy = STRATEGIES[strategy].new(settings, devicegraph)
  else
    err_msg = "Unsupported device planner strategy :#{strategy}"
    log.error err_msg
    raise ArgumentError, err_msg 
  end
end

```